### PR TITLE
feat: grant dormancy check

### PR DESCRIPTION
**Pull request by rahmatrhd**

### CODE REVIEW

File: jobs/grant_dormancy_check.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		return fmt.Errorf("invalid retain grant for duration: %w", err)
+	}
+
+	providers, err := h.providerService.Find(ctx)
+	if err != nil {
+		return fmt.Errorf("listing providers: %w", err)
+	}
+
+	for _, p := range providers {
+		if p.URN != "pilotdata-integration-bigquery" {
```
Comment (bsushmith) [inline jobs/grant_dormancy_check.go:38]: Why is this needed here?

File: jobs/grant_dormancy_check.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"fmt"
+	"time"
+
+	"github.com/goto/guardian/domain"
+)
+
+type GrantDormancyCheckConfig struct {
+	DryRun         bool   `mapstructure:"dry_run"`
+	Period         string `mapstructure:"period"`
+	RetainGrantFor string `mapstructure:"retain_grant_for"`
```
Comment (bsushmith) [inline jobs/grant_dormancy_check.go:14]: Can you add comment to both the fields for sample duration values.

File: core/grant/service.go [hunk 3]
```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
...
+		}
+	}
+
+	s.logger.Info(fmt.Sprintf("getting activities for provider %q", provider.URN))
+	activities, err := s.providerService.ListActivities(ctx, *provider, domain.ListActivitiesFilter{
+		AccountIDs:   uniqueAccountIDs,
+		TimestampGte: &startDate,
+	})
+	if err != nil {
+		return fmt.Errorf("listing activities: %w", err)
```
Comment (bsushmith) [inline core/grant/service.go:530]: can you add provider id in this error message

File: core/grant/service.go [hunk 4]
```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
...
+					},
+				},
+			})
+		}
+
+		if errs := s.notifier.Notify(notifications); errs != nil {
+			for _, err1 := range errs {
+				s.logger.Error("failed to send notifications", "error", err1.Error())
+			}
+		}
```
Comment (bsushmith) [inline core/grant/service.go:607]: Are there test cases for this functionality?

File: core/grant/service.go [hunk 5]
```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
...
+	prepare_notifications:
+		for owner, grants := range dormantGrantsByOwner {
+			var grantsMap []map[string]interface{}
+			var grantIDs []string
+
+			for _, g := range grants {
+				grantMap, err := structToMap(g)
+				if err != nil {
+					s.logger.Error("failed to convert grant to map", "error", err)
+					continue prepare_notifications
```
Comment (bsushmith) [inline core/grant/service.go:580]: is this needed?

File: core/grant/service.go [hunk 6]
```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
...
+				},
+				Message: domain.NotificationMessage{
+					Type: domain.NotificationTypeUnusedGrant,
+					Variables: map[string]interface{}{
+						"dormant_grants":       grantsMap,
+						"period":               criteria.Period.String(),
+						"retain_duration":      criteria.RetainDuration.String(),
+						"start_date_formatted": startDate.Format("Jan 02, 2006 15:04:05 UTC"),
+					},
+				},
```
Comment (bsushmith) [inline core/grant/service.go:599]: Can you put down a sample notification here?

File: core/grant/service.go [hunk 7]
```diff
@@ -533,3 +663,28 @@ func reduceGrantsByProviderRole(rc domain.ResourceConfig, grants []*domain.Grant
...
+			return nil, err
+		}
+
+		if err := json.Unmarshal(jsonString, &result); err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
```
Comment (bsushmith) [inline core/grant/service.go:690]: this same functionality is also here - https://github.com/goto/guardian/blob/main/domain/appeal.go#L264. Can we keep this in utils and reuse it.

File: domain/activity.go [hunk 8]
```diff
@@ -74,3 +75,39 @@ func (f *ImportActivitiesFilter) GetResources() []*Resource {
 
 	return resources
 }
+
+// act exists in domain and used across services
+// this should be associatable with grants
+type act struct {
```
Comment (bsushmith) [inline domain/activity.go:81]: this is not used anywhere

File: domain/activity.go [hunk 9]
```diff
@@ -74,3 +75,39 @@ func (f *ImportActivitiesFilter) GetResources() []*Resource {
...
+
+// bigqueryAct exists in plugin
+type bigqueryAct struct {
+	AccountID    string
+	ResourceName string   // --> resourceURN, resourceType
+	Permissions  []string // gcp permissions
+	Timestamp    time.Time
+}
+
+type grants []grants
```
Comment (bsushmith) [inline domain/activity.go:112]: can you run `make lint` in local and resolve the `unused` errors.

File: plugins/notifiers/slack/templates/UnusedGrant.json [hunk 10]
```diff
@@ -0,0 +1,9 @@
+[
+  {
+    "type":"section",
+    "text":{
+      "type":"mrkdwn",
+      "text":"We have advanced the expiration date for the following grants due to inactivity since `{{.start_date}}`:{{ range .dormant_grants }}\n>*ID*: `{{.id}}`\n>*Account ID*: `{{.account_id}}`\n>*Resource*: `{{.resource.urn}}` ({{.resource.provider_type}} {{.resource.type}})\n>*Role*: `{{.role}}`\n>*Expiration Date* (new): `{{.expiration_date}}`\n{{end}}"
```
Comment (bsushmith) [inline plugins/notifiers/slack/templates/UnusedGrant.json:6]: Can you add a sample formatted output of this text?

File: plugins/providers/bigquery/activity.go [hunk 11]
```diff
@@ -8,10 +8,8 @@ import (
 	"strings"
 	"time"
 
-	"cloud.google.com/go/logging"
-	"cloud.google.com/go/logging/logadmin"
 	"github.com/goto/guardian/domain"
-	"google.golang.org/api/iterator"
+	"google.golang.org/api/logging/v2"
```
Comment (bsushmith) [inline plugins/providers/bigquery/activity.go:12]: Any specific reason we are changing the library being used?

File: plugins/providers/bigquery/model.go [hunk 12]
```diff
@@ -146,3 +170,31 @@ func (r BigQueryResourceName) BigQueryResourceID() string {
...
+type cloudLoggingOptions struct {
+	LogBucket string `json:"log_bucket" yaml:"log_bucket" mapstructure:"log_bucket"`
+}
+
+type activityConfig struct {
+	*domain.ActivityConfig
+}
+
+func (c activityConfig) Validate() error {
+	// TODO:
```
Comment (bsushmith) [inline plugins/providers/bigquery/model.go:183]: todo

File: plugins/providers/bigquery/model.go [hunk 13]
```diff
@@ -7,15 +7,39 @@ import (
...
+	u = strings.TrimSpace(u)
+	items := strings.Split(u, " ")
+
+	if len(items) == 2 {
+		return ResourceTypeDataset
+	} else if len(items) == 3 {
+		return ResourceTypeTable
+	}
+	return resourceTypeUnknown
+}
```
Comment (bsushmith) [inline plugins/providers/bigquery/model.go:41]: can you add test cases for this snippet

File: plugins/providers/bigquery/provider.go [hunk 14]
```diff
@@ -340,6 +374,120 @@ func (p *Provider) GetActivities(ctx context.Context, pd domain.Provider, filter
...
+		return nil, fmt.Errorf("invalid bucket's retention period: %q: %w", logBucket.RetentionDays, err)
+	}
+	if filter.TimestampGte != nil && time.Since(*filter.TimestampGte) > retentionDuration {
+		return nil, fmt.Errorf("%w: log bucket's retention in days: %q", ErrActivityLogRetentionPeriodExceeded, logBucket.RetentionDays)
+	} else {
+		t := time.Now().Add(-retentionDuration)
+		filter.TimestampGte = &t
+	}
+
+	// TODO: check private log viewer access is granted
```
Comment (bsushmith) [inline plugins/providers/bigquery/provider.go:416]: todo

File: plugins/providers/bigquery/provider.go [hunk 15]
```diff
@@ -497,3 +645,17 @@ func translateDatasetRoleToBigQueryRole(role string) string {
...
+	for _, element := range subset {
+		checkset[element] = true
+	}
+	for _, element := range superset {
+		if checkset[element] {
+			delete(checkset, element)
+		}
+	}
+	return len(checkset) == 0
+}
```
Comment (bsushmith) [inline plugins/providers/bigquery/provider.go:661]: test cases for this module

File: core/grant/service.go [hunk 16]
```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
...
+
+	s.logger.Info("checking grants dormancy...")
+	var dormantGrants []*domain.Grant
+	var dormantGrantsIDs []string
+	var dormantGrantsByOwner = map[string][]*domain.Grant{}
+	for _, g := range grantsPointer {
+		if len(g.Activities) == 0 {
+			g.ExpirationDateReason = fmt.Sprintf("%s: %s", domain.GrantExpirationReasonDormant, criteria.RetainDuration)
+			newExpDate := time.Now().Add(criteria.RetainDuration)
+			g.ExpirationDate = &newExpDate
```
Comment (bsushmith) [inline core/grant/service.go:550]: Can we keep this expiration date intact as it is as its based on user input, and create another column explicitly for this functionality? 
and in the grant expiration reminder and revocation job, the new field can be included for expiry.

**Review by bsushmith — CHANGES_REQUESTED** [review]

File: api/handler/v1beta1/activity.go [hunk 17]
```diff
@@ -66,7 +66,7 @@ func (s *GRPCServer) ListActivities(ctx context.Context, req *guardianv1beta1.Li
 }
 
 func (s *GRPCServer) ImportActivities(ctx context.Context, req *guardianv1beta1.ImportActivitiesRequest) (*guardianv1beta1.ImportActivitiesResponse, error) {
-	filter := domain.ImportActivitiesFilter{
+	filter := domain.ListActivitiesFilter{
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply api/handler/v1beta1/activity.go:69]: correct

File: cli/job.go [hunk 18]
```diff
@@ -43,13 +43,15 @@ func runJobCmd() *cobra.Command {
...
 			$ guardian job run revoke_grants_by_user_criteria
+			$ guardian job run grant_dormancy_check
 		`),
 		Args: cobra.ExactValidArgs(1),
 		ValidArgs: []string{
 			string(jobs.TypeFetchResources),
 			string(jobs.TypeExpiringGrantNotification),
 			string(jobs.TypeRevokeExpiredGrants),
 			string(jobs.TypeRevokeGrantsByUserCriteria),
+			string(jobs.TypeGrantDormancyCheck),
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply cli/job.go:54]: without string conversion it's throwing error:
`cannot use jobs.TypeGrantDormancyCheck (constant "grant_dormancy_check" of type jobs.Type) as string value`

File: core/grant/service.go [hunk 19]
```diff
@@ -480,6 +484,132 @@ func (s *Service) ImportFromProvider(ctx context.Context, criteria ImportFromPro
...
+	if criteria.DryRun {
+		s.logger.Info("dry run mode, skipping updating grants expiration date")
+		return nil
+	}
+
+	if err := s.repo.BulkUpsert(ctx, dormantGrants); err != nil {
+		return fmt.Errorf("updating grants expiration date: %w", err)
+	}
+
+	go func() { // send notifications
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply core/grant/service.go:569]: regardless sending notification is successful or error, we would still return `nil` error, so the function's return is not depending on it. Running in a goroutine just completely isolating and showing that it is independent and non-blocking

File: core/grant/service.go (hunk 5 above)
Comment (rahmatrhd) [reply core/grant/service.go:580]: this was to avoid making the entire notifications not being sent if one parsing errors

File: jobs/grant_dormancy_check.go (hunk 1 above)
Comment (rahmatrhd) [reply jobs/grant_dormancy_check.go:38]: this was for testing, will remove it 🙊

File: plugins/providers/bigquery/activity.go (hunk 11 above)
Comment (rahmatrhd) [reply plugins/providers/bigquery/activity.go:12]: we now use getLoggingBucket API which was not supported in `cloud.google.com/go/logging`, so rather than using two different clients I replace the other calls to use this `logging/v2` instead.

File: cli/job.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (bsushmith) [reply cli/job.go:54]: weird. I expected this to work. But sure lets leave it as it is.

File: core/grant/service.go (hunk 16 above)
Comment (rahmatrhd) [reply core/grant/service.go:547]: we have introduced `requested_expiration_date` in #27 which will record the initial expiration date as requested in the appeal and this field will never be updated.

File: core/grant/service.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (mabdh) [reply core/grant/service.go:566]: Have we tested this go routine behaviour? 

I am afraid there will be some unwanted failure or behaviour since the preparation notification inside goroutine require some information from outside goroutine.

If there is a problem, maybe we could just put the `s.notifier.Notify` function only in goroutine and keep the message preparation outside.

File: core/policy/service.go [hunk 20]
```diff
@@ -217,7 +217,7 @@ func (s *Service) decryptAndDeserializeIAMConfig(c *domain.IAMConfig) error {
 	if err := iamClientConfig.Decrypt(); err != nil {
 		return fmt.Errorf("decrypting iam config: %w", err)
 	}
-	iamClientConfigMap, err := structToMap(iamClientConfig)
+	iamClientConfigMap, err := utils.StructToMap(iamClientConfig)
```
Comment (mabdh) [inline core/policy/service.go:220]: rather than having `utils` package (which should be avoided in golang). why not adding a function receiver of the config, so converting the config to map would be something like:
`iamClientConfig.ToMap()`

Comment (rahmatrhd) [reply core/policy/service.go:220]: this is for avoiding redundancy of writing the same logic multiple times. In that case, we would implement the same logic for `x.ToMap()`, `y.ToMap()`, `z.ToMap()` and so on, and then when we want to refactor it, it would go back to the approach where we use the same shared function (`utils.StructToMap()`) all over again

File: core/grant/service.go (hunk 3 above)
Comment (rahmatrhd) [reply core/grant/service.go]: added provider.URN

File: core/policy/service.go (hunk 20 above)
Comment (rahmatrhd) [reply core/policy/service.go:220]: IMO it's still acceptable to have this shared function approach since the logics are exactly the same and not related with any domain or business context, it's purely a helper function. Based on what I see on existing packages for similar use cases, we can move this helper to `pkg/structs` instead, following golang's built-in `strings` helper fashion. WDYT?

there are 4 components (including 1 in a test package) used with `utils.StructToMap`

Comment (mabdh) [reply core/policy/service.go:220]: yes we can do that later

File: plugins/providers/bigquery/provider.go (hunk 15 above)
Comment (rahmatrhd) [reply plugins/providers/bigquery/provider.go:676]: updated

File: plugins/providers/bigquery/model.go (hunk 13 above)
Comment (rahmatrhd) [reply plugins/providers/bigquery/model.go]: removed, this is unused

File: core/grant/service.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply core/grant/service.go:567]: I see, I will just make it synchronous and test it out for now to avoid unwanted failure, since this part is also a little bit expensive to test

**Review by mabdh — APPROVED** [review]

File: core/grant/service.go (hunk 4 above)
Comment (rahmatrhd) [reply core/grant/service.go]: test cases added
